src/bsoncxx/lib/bsoncxx/v1/element/view.cpp

```diff
@@ -12,19 +12,337 @@
 
+class alignas(BSONCXX_PRIVATE_MAX_ALIGN_T) view::impl {
+   private:
+    enum : std::size_t {
+        _padding_size = sizeof(view::_storage)       // Total reserved.
+                        - sizeof(void*)              // _raw
+                        - 3u * sizeof(std::uint32_t) // _length, _offset, and _keylen.
+                        - 1u                         // _is_valid (final byte).
+                        - 0u,
+    };
+
+    // `_padding_size == 3` given `sizeof(void*) == 8`.
+    static_assert(_padding_size < sizeof(view::_storage), "sizeof(impl) must not exceed reserved storage size");
+
+    std::uint8_t const* _raw = {};
+    std::uint32_t _length = {};
+    std::uint32_t _offset = {};
+    std::uint32_t _keylen = {};
+
+    BSONCXX_PRIVATE_WARNINGS_PUSH();
+    BSONCXX_PRIVATE_WARNINGS_DISABLE(GNU("-Wunused"));
+    unsigned char _padding[_padding_size]; // Reserved.
+    BSONCXX_PRIVATE_WARNINGS_POP();
+
+    bool _is_valid = {}; // Last byte.
+
+   public:
+    impl() = default;
+
+    impl(
+        std::uint8_t const* raw,
+        std::uint32_t length,
+        std::uint32_t offset,
+        std::uint32_t keylen,
+        bool is_valid = true)
+        : _raw{raw}, _length{length}, _offset{offset}, _keylen{keylen}, _is_valid{is_valid} {}
+
+    void check() const;
+
+    std::uint8_t const* raw() const {
+        return _raw;
+    }
+
+    std::uint32_t length() const {
+        return _length;
+    }
+
+    std::uint32_t offset() const {
+        return _offset;
+    }
+
+    std::uint32_t keylen() const {
+        return _keylen;
+    }
+
+    bool is_valid() const {
+        return _raw && _is_valid;
+    }
+
+    impl to_invalid() const {
+        auto ret = *this;
+        ret._is_valid = false;
+        return ret;
+    }
+
+    v1::types::id type_id() const {
+        auto const iter = this->iter();
+        return static_cast<v1::types::id>(bson_iter_type(&iter));
+    }
+
+    v1::types::id type_id_unchecked() const {
+        if (auto const iter_opt = this->iter_unchecked()) {
+            return static_cast<v1::types::id>(bson_iter_type(&*iter_opt));
+        }
+        return v1::types::id{}; // BSON_TYPE_EOD
+    }
+
+    v1::stdx::optional<bson_iter_t> iter_unchecked() const {
+        bson_iter_t iter;
+        if (bson_iter_init_from_data_at_offset(&iter, _raw, _length, _offset, _keylen)) {
+            return iter;
+        }
+        return v1::stdx::nullopt;
+    }
+
+    bson_iter_t iter() const {
+        this->check();
+        if (auto iter_opt = this->iter_unchecked()) {
+            return *iter_opt;
+        }
+        throw v1::exception{code::invalid_data};
+    }
+
+    v1::types::view type_view() const {
+        this->check();
+        return this->type_view_unchecked();
+    }
+
+    v1::types::view type_view_unchecked() const {
+        if (auto opt = v1::types::view::internal::make(_raw, _length, _offset, _keylen)) {
```

Comment:
This invokes the internal function `bsoncxx::v1::view::internal::make(bson_value_t const& v)` in `lib/bsoncxx/v1/types/view.cpp`.